families/supplychain/python/sawtooth_supplychain/rest_api/route_handlers.py

```diff
@@ -0,0 +1,756 @@
+# Copyright 2017 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+# Supply Chain REST API
+
+import logging
+import json
+import re
+from concurrent.futures import ThreadPoolExecutor
+import psycopg2
+import aiopg
+from aiohttp import web
+
+import sawtooth_supplychain.rest_api.exceptions as errors
+
+
+DEFAULT_TIMEOUT = 300
+LOGGER = logging.getLogger(__name__)
+DEFAULT_PAGE_SIZE = 1000    # Reasonable default
+
+
+async def db_query(db_cnx, query, query_tuple):
+    try:
+        async with aiopg.create_pool(db_cnx) as pool:
+            async with pool.acquire() as conn:
+                async with conn.cursor() as cur:
+                    # Execute the query
+                    try:
+                        await cur.execute(query, query_tuple)
+                    except:
+                        LOGGER.exception("Could not execute query: %s", query)
+                        raise errors.UnknownDatabaseError()
+
+                    # Fetch the rows
+                    rows = []
+                    async for row in cur:
+                        rows.append(row)
+                    return rows
+
+    except psycopg2.OperationalError as e:
+        raise e
+
+
+class RouteHandler(object):
+    """Contains a number of aiohttp handlers for endpoints in
+    the Supply Chain Rest Api.
+
+    Each handler takes an aiohttp Request object, and uses the data in
+    that request to create queries to send to the database. The query
+    results are processed, and finally an aiohttp Response object is
+    sent back to the client with JSON formatted data and metadata.
+
+    If something goes wrong, an aiohttp HTTP exception is raised or returned
```

Comment:
This docstring line says "an aiohttp HTTP exception is raised **or returned**". If a route handler `return`s an HTTPError, aiohttp will recognize it, and send back an error code to the client correctly. It works identically to `raise`ing the error.

At some point my REST API used both behaviors, and that is when I wrote that line. It is probably inaccurate for the Sawtooth REST API at this point, but it is definitely inaccurate for this REST API, since you never `return` an error.